### PR TITLE
Support Add Sub Mul Max Min Pow binary functors in elementwise system

### PR DESCRIPTION
**Pull request by JamesLim-sy**
### PR types

Performance optimization

### PR changes

OPs

### Describe

1. Basing on new elementwise + broadcast system， support binary functors below : 
    Add Sub Mul Max Min Pow

2. The performance Optimizaiton can be shown in table below:
截屏2021-06-01 下午6 47 19

3. As can be seen from the table, the performance of supported ops  has improved a lot in most cases and catches up with Pytorch after adopting new elementwise + broadcast system. However,  the old broadcast subsystem works better in the 2nd test case, the old broadcast subsystem consists of perf-optimized  branch and common broadcast branch, the former one works well when the quantity of input tensor data is not big enough and the input tensor`s dim meet the special demands. Apparently,  2nd test case perfectly meets the that demands and  data quantity  is relatively small, therefore, it beats the new elementwise + broadcast system in 2nd case. But introducing of this branch in old broadcast-elementwise dose make the code less compactness and hard to maintain, and  working area of it is not big enough.  Furthermore, the elementwise op in paddle is suggested to dealing with  NN whose data quantity is often large,  so i suggest Approval of 2nd test case and adopt the new elementwise + broadcast system in these binary op.

### CODE REVIEW

File: paddle/fluid/operators/elementwise/elementwise_max_op.cu [hunk 1]
```diff
@@ -12,9 +12,46 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+class ElementwiseMaxKernel<platform::CUDADeviceContext, T>
+    : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* x = ctx.Input<framework::LoDTensor>("X");
+    auto* y = ctx.Input<framework::LoDTensor>("Y");
+    auto* z = ctx.Output<framework::LoDTensor>("Out");
+    z->mutable_data<T>(ctx.GetPlace());
+    int axis = ctx.Attr<int>("axis");
+    axis = axis == -1 ? std::abs(x->dims().size() - y->dims().size()) : axis;
```
Comment (Xreki) [inline paddle/fluid/operators/elementwise/elementwise_max_op.cu]: 这行axis的换算也可以放到`LaunchElementwiseCudaKernel`里面？

File: paddle/fluid/operators/elementwise/elementwise_max_op.h [hunk 2]
```diff
@@ -35,8 +35,8 @@ class ElementwiseMaxKernel : public framework::OpKernel<T> {
 
     z->mutable_data<T>(ctx.GetPlace());
     int axis = ctx.Attr<int>("axis");
-    ElementwiseComputeEx<MaxFunctor<T>, DeviceContext, T>(ctx, x, y, axis,
-                                                          MaxFunctor<T>(), z);
+    ElementwiseComputeEx<MaxFunctor<T>, platform::CPUDeviceContext, T>(
```
Comment (Xreki) [inline paddle/fluid/operators/elementwise/elementwise_max_op.h]: 这是一个通用的实现，不是一个特化的实现，不要改这里。后面如果想删除`ElementwiseComputeEx`的GPU实现代码，可以把`ElementwiseMaxKernel`在.h里面声明，CPU Kernel在.cc里面特化，GPU Kernel在.cu里面特化。

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu [hunk 3]
```diff
@@ -25,37 +26,60 @@ namespace paddle {
...
-        grid_size, block_size, 0,
-        ctx.template device_context<platform::CUDADeviceContext>().stream()>>>(
-        x2, y2, z2, size);
+template <typename T>
+class ElementwiseMulKernel<platform::CUDADeviceContext, T>
+    : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto x_var = ctx.InputVar("X");
+    PADDLE_ENFORCE_EQ(x_var != nullptr, true,
```
Comment (Xreki) [inline paddle/fluid/operators/elementwise/elementwise_mul_op.cu]: 这个检查可以用`PADDLE_ENFORCE_NOT_NULL`。

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu [hunk 4]
```diff
@@ -25,37 +26,60 @@ namespace paddle {
...
+      z = ctx.Output<framework::SelectedRows>("Out")->mutable_value();
+    } else if (x_var->IsType<framework::LoDTensor>()) {
+      x = x_var->Get<framework::LoDTensor>();
+      z = ctx.Output<framework::LoDTensor>("Out");
+    } else {
+      PADDLE_THROW(platform::errors::InvalidArgument(
+          "X's type[%s] is not supported by elementwise_op. X's type should be "
+          "LoDTensor or SelectedRows.",
+          framework::ToTypeName(x_var->Type())));
+    }
```
Comment (Xreki) [inline paddle/fluid/operators/elementwise/elementwise_mul_op.cu:83]: 不要复制粘贴大段代码，L41 - L70写个函数封装一下。

File: paddle/fluid/operators/elementwise/elementwise_max_op.h (hunk 2 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/elementwise/elementwise_max_op.h]: 按照建议修改

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu (hunk 3 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/elementwise/elementwise_mul_op.cu]: 按照建议修改

File: paddle/fluid/operators/elementwise/elementwise_op_broadcast.cu.h [hunk 5]
```diff
@@ -501,23 +502,30 @@ void LaunchBroadcastElementwiseCudaKernel(
   }
 }
 
-template <ElementwiseType ET, typename InT, typename OutType, typename Functor>
+template <ElementwiseType ET, typename InT, typename OutT, typename Functor>
 void LaunchElementwiseCudaKernel(
-    const platform::CUDADeviceContext &cuda_ctx,
+    const framework::ExecutionContext &ctx,
```
Comment (Xreki) [inline paddle/fluid/operators/elementwise/elementwise_op_broadcast.cu.h]: 传`platform::CUDADeviceContext`，不要修改这个参数类型，不要include operator.h

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu (hunk 4 above)
Comment (Xreki) [reply paddle/fluid/operators/elementwise/elementwise_mul_op.cu:83]: L41 - L70也可以封装到`PackTensorsIntoVector`函数里面。

**Review by Xreki — APPROVED** [review]
LGTM. 还有些代码层面的优化，可以下个PR再做。

**Pull request merged**